File: dotnet/src/Envdoctor/Detectors/SchemaValidationDetector.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using Envdoctor.Config;
using Envdoctor.Models;
using static Envdoctor.Detectors.DetectorHelpers;

namespace Envdoctor.Detectors;

/// Schema-validation: a variable value does not match its declared schema.
public sealed class SchemaValidationDetector : IDetector
{
    public string Id => "schema-validation";
    public string Name => "schema-validation";
    public string Description => "A variable value does not match its declared schema.";

    private delegate string? Validator(string value);

    /// Build a validator for a variable schema; returns null on success.
    private static Validator? BuildValidator(VariableSchema schema)
    {
        if (schema.EnumValues is { Count: > 0 } enumValues)
        {
            var values = enumValues.ToList();
            return v => values.Contains(v) ? null : $"must be one of: {string.Join(", ", values)}";
        }

        switch (schema.VarType)
        {
            case SchemaType.Integer:
            {
                var min = schema.Min;
                var max = schema.Max;
                return v =>
                {
                    if (!long.TryParse(v, System.Globalization.NumberStyles.Integer,
                            System.Globalization.CultureInfo.InvariantCulture, out var n))
                        return "must be an integer";
                    if (min is not null && n < min.Value)
                        return $"must be >= {min.Value}";
                    if (max is not null && n > max.Value)
                        return $"must be <= {max.Value}";
                    return null;
                };
            }
            case SchemaType.Float:
            {
                var min = schema.Min;
                var max = schema.Max;
                return v =>
                {
                    if (!double.TryParse(v, System.Globalization.NumberStyles.Float,
                            System.Globalization.CultureInfo.InvariantCulture, out var n))
                        return "must be a float";
                    if (min is not null && n < min.Value)
                        return $"must be >= {min.Value}";
                    if (max is not null && n > max.Value)
                        return $"must be <= {max.Value}";
                    return null;
                };
            }
            case SchemaType.Boolean:
                return v => v is "true" or "false" ? null : "must be a boolean";
            case SchemaType.Url:
                return v => v.StartsWith("http://", StringComparison.Ordinal) ||
                            v.StartsWith("https://", StringComparison.Ordinal)
                    ? null
                    : "must be a valid URL";
            case SchemaType.Json:
                return v =>
                {
                    try
                    {
                        JsonDocument.Parse(v);
                        return null;
                    }
                    catch (JsonException)
                    {
                        return "must be valid JSON";
                    }
                };
            case SchemaType.Regex:
            {
                if (schema.Regex is not { } regexStr)
                    return null;
                Regex re;
                try
                {
                    re = new Regex(regexStr);
                }
                catch (ArgumentException)
                {
                    return null;
                }
                return v => re.IsMatch(v) ? null : $"must match {regexStr}";
            }
            case SchemaType.String:
            case SchemaType.Enum:
            case null:
                return _ => null;
            default:
                return _ => null;
        }
    }

    private static string? ValidateValue(string? value, VariableSchema schema)
    {
        if (value is null || value.Trim().Length == 0)
            return schema.Optional == true ? null : "value is required";

        var validator = BuildValidator(schema);
        return validator?.Invoke(value);
    }

    public List<Finding> Detect(IndexedModel index)
    {
        var schema = index.Model.Config.Schema;
        if (schema.Count == 0)
            return new List<Finding>();

        var findings = new List<Finding>();

        var entries = index.EnvDefinitions
            .OrderBy(kv => DefSortKey(kv.Value), Comparer<(string Path, int Line)>.Create(CompareKeys))
            .ThenBy(kv => kv.Key, StringComparer.Ordinal);

        foreach (var (name, defs) in entries)
        {
            if (!schema.TryGetValue(name, out var variableSchema))
                continue;

            foreach (var def in defs)
            {
                var error = ValidateValue(def.Value, variableSchema);
                if (error is null)
                    continue;
                findings.Add(MakeFinding(
                    "schema-validation",
                    Severity.Error,
                    name,
                    $"does not match schema: {error}",
                    new List<Origin> { def.Origin.Clone() }));
            }
        }

        return findings;
    }
}

File: swift/Sources/EnvdoctorKit/Detectors/SchemaValidationDetector.swift
import Foundation

/// Schema validation: values in env files are checked against the per-variable
/// rules declared in the config's `schema` section.
public struct SchemaValidationDetector: Detector {
    public let id = "schema-validation"
    public let name = "schema-validation"
    public let description = "A variable value does not match its declared schema."

    public init() {}

    public func detect(_ index: IndexedModel) -> [Finding] {
        let schema = index.model.config.schema
        if schema.isEmpty { return [] }

        // Deterministic emission order: earliest (file, line) of each name's
        // definitions, then name — matching the other ports.
        let sortedEntries = index.envDefinitions.sorted { a, b in
            let ka = sortKey(a.1)
            let kb = sortKey(b.1)
            if ka.path != kb.path { return ka.path < kb.path }
            if ka.line != kb.line { return ka.line < kb.line }
            return a.0 < b.0
        }

        var findings: [Finding] = []
        for (name, defs) in sortedEntries {
            guard let variableSchema = schema[name] else { continue }

            for def in defs {
                guard let error = validateValue(def.value, variableSchema) else { continue }
                findings.append(makeFinding(
                    "schema-validation", .error, name,
                    "does not match schema: \(error)",
                    [def.origin]
                ))
            }
        }
        return findings
    }

    private func sortKey(_ defs: [Definition]) -> (path: String, line: Int) {
        var best: (String, Int)?
        for def in defs {
            let candidate = (def.origin.filePath, def.origin.line ?? Int.max)
            if let current = best {
                let better = candidate.0 == current.0
                    ? candidate.1 < current.1
                    : candidate.0 < current.0
                if better { best = candidate }
            } else {
                best = candidate
            }
        }
        guard let first = best else { return ("", 0) }
        return first
    }

    /// Returns nil when the value is valid, otherwise the error message.
    private func validateValue(_ value: String?, _ schema: VariableSchema) -> String? {
        guard let value, !value.trimmingCharacters(in: .whitespaces).isEmpty else {
            return (schema.optional ?? false) ? nil : "value is required"
        }

        if let enumValues = schema.enumValues, !enumValues.isEmpty {
            return enumValues.contains(value)
                ? nil
                : "must be one of: \(enumValues.joined(separator: ", "))"
        }

        switch schema.type {
        case .integer:
            guard let n = Int64(value.trimmingCharacters(in: .whitespaces)) else {
                return "must be an integer"
            }
            if let min = schema.min, Double(n) < min { return "must be >= \(Json.jsNumber(min))" }
            if let max = schema.max, Double(n) > max { return "must be <= \(Json.jsNumber(max))" }
            return nil

        case .float:
            guard let n = Double(value.trimmingCharacters(in: .whitespaces)) else {
                return "must be a float"
            }
            if let min = schema.min, n < min { return "must be >= \(Json.jsNumber(min))" }
            if let max = schema.max, n > max { return "must be <= \(Json.jsNumber(max))" }
            return nil

        case .boolean:
            return (value == "true" || value == "false") ? nil : "must be a boolean"

        case .url:
            return (value.hasPrefix("http://") || value.hasPrefix("https://"))
                ? nil
                : "must be a valid URL"

        case .json:
            return (try? JsonParser.parse(value)) != nil ? nil : "must be valid JSON"

        case .regex:
            guard let pattern = schema.regex,
                  let re = try? NSRegularExpression(pattern: pattern) else {
                return nil
            }
            let range = NSRange(value.startIndex..<value.endIndex, in: value)
            return re.firstMatch(in: value, range: range) != nil ? nil : "must match \(pattern)"

        case .string, .enum, .none:
            return nil
        }
    }
}
